File: components/definitions/audio_settings.py
from bpy.props import FloatProperty, EnumProperty
from .hubs_component import HubsComponent
from ..types import Category, PanelType, NodeType
from ..consts import DISTACE_MODELS

# TODO Add this component in the scene by default?


class hubs_component_audio_settings(HubsComponent):
    _definition = {
        'export_name': 'audio-settings',
        'display_name': 'Audio Settings',
        'category': Category.SCENE,
        'node_type': NodeType.SCENE,
        'panel_type': PanelType.SCENE
    }

    avatarDistanceModel: EnumProperty(
        name="Avatar Distance Model",
        description="Avatar Distance Model",
        items=DISTACE_MODELS,
        default="inverse")

    avatarRolloffFactor: FloatProperty(
        name="Avatar Rolloff Factor", default=2.0, min=0.0, soft_min=0.0)

    avatarRefDistance: FloatProperty(
        name="Avatar Ref Distance", description="Avatar Ref Distance", subtype="DISTANCE", default=1.0, min=0.0, soft_min=0.0)

    avatarMaxDistance: FloatProperty(
        name="Avatar Max Distance", description="Avatar Max Distance", subtype="DISTANCE", default=10000.0, min=0.0, soft_min=0.0)

    mediaVolume: FloatProperty(
        name="Media Volume", description="Media Volume", default=0.5, min=0.0, soft_min=0.0)

    mediaDistanceModel: EnumProperty(
        name="Media Distance Model",
        description="Media Distance Model",
        items=DISTACE_MODELS,
        default="inverse")

    mediaRolloffFactor: FloatProperty(
        name="Media Rolloff Factor", description="Media Rolloff Factor", default=2.0, min=0.0, soft_min=0.0)

    mediaRefDistance: FloatProperty(
        name="Media Ref Distance", description="Media Rolloff Factor", subtype="DISTANCE", default=2.0, min=0.0, soft_min=0.0)

    mediaMaxDistance: FloatProperty(
        name="Media Max Distance", description="Media Max Distance", subtype="DISTANCE", default=10000.0, min=0.0, soft_min=0.0)

    mediaConeInnerAngle: FloatProperty(
        name="Media Cone Inner Angle", description="Media Cone Inner Angle", default=360.0, min=0.0, soft_min=0.0, max=360.0, soft_max=360.0)

    mediaConeOuterAngle: FloatProperty(
        name="Media Cone Inner Angle", description="Media Cone Outer Angle", default=0.0, min=0.0, soft_min=0.0, max=360.0, soft_max=360.0)

    mediaConeOuterGain: FloatProperty(
        name="Media Cone Inner Angle", description="Media Cone Outer Gain", default=0.0, min=0.0, soft_min=0.0)

File: tests/test/tests/audio-settings.js
const fs = require('fs');
const path = require('path')
const assert = require('assert');
const utils = require('../utils.js');

module.exports = {
    description: 'can export audio-settings',
    test: outDirPath => {
        let gltfPath = path.resolve(outDirPath, 'audio-settings.gltf');
        const asset = JSON.parse(fs.readFileSync(gltfPath));

        assert.strictEqual(asset.extensionsUsed.includes('MOZ_hubs_components'), true);
        assert.strictEqual(utils.checkExtensionAdded(asset, 'MOZ_hubs_components'), true);

        const scene = asset.scenes[0];
        assert.strictEqual(utils.checkExtensionAdded(scene, 'MOZ_hubs_components'), true);

        const ext = scene.extensions['MOZ_hubs_components'];
        assert.deepStrictEqual(ext, {
            "audio-settings": {
                "avatarDistanceModel": "inverse",
                "avatarRolloffFactor": 2,
                "avatarRefDistance": 1,
                "avatarMaxDistance": 10000,
                "mediaVolume": 0.5,
                "mediaDistanceModel": "inverse",
                "mediaRolloffFactor": 2,
                "mediaRefDistance": 1,
                "mediaMaxDistance": 10000,
                "mediaConeInnerAngle": 360,
                "mediaConeOuterAngle": 0,
                "mediaConeOuterGain": 0
            }
        });
    }
};